### ISSUE TITLE
Psalter must check `@inheritdoc` tag

### ISSUE THREAD

**Issue by WinterSilence** [vimeo/psalm#6572]
Source code:
~~~php
class A
{
    /**
     * @var string Base Url
     */
    public $baseUrl = '';
}
class B extends A
{
    /**
     * @inheritDoc
     */
    public $baseUrl = '...';
}
~~~
After `psalter --issues=all`, expected result - no changes, current result:
~~~php
class A
{
    /**
     * @var string Base Url
     * @psalm-var ''
     */
    public $baseUrl = '';
}
class B extends A
{
    /**
     * @inheritDoc
     * @var string
     * @psalm-var '...'
     */
    public $baseUrl = '...';
}
~~~

**Comment by WinterSilence** [comment]
https://psalm.dev/r/107f171a6e

**Comment by orklah** [comment]
Can you explain why @inheritDoc would change something here?

**Comment by WinterSilence** [comment]
@orklah Because `@inheritdoc` inherit parent tags. Psalm append redundant tags - it's can be optional.

**Comment by orklah** [comment]
Oh, I see. Indeed `@var string` is redundant, but the others, less so.

Psalm currently have no type inheritance between properties. This is related to #6504 and #6575. I'm not sure if this was done on purpose initially however...

**Comment by WinterSilence** [comment]
@orklah why you think `@psalm-var ''` is not redundant? where is "profit"?

**Comment by orklah** [comment]
Ok, I checked a little what Psalm is doing

It sees you don't have a type in signature so it starts trying to fix the issue.

When it fixes the issue, Psalm will try to infer the best possible type, and then try to express it the best it can, either through the signature, the standard phpdoc or the specific psalm phpdoc.

There seem to be an oversight in case a property is inherited because it doesn't see there is a default value(if we add a __construct, it will stop trying to change the parent type: https://psalm.dev/r/9911c84394). So from its perpective, your property can only have the value `''` and it express it as such.

**Comment by WinterSilence** [comment]
@orklah it's strange behavior for **public** properties

**Comment by orklah** [comment]
I agree on that, it's a legit bug: https://psalm.dev/r/660a502141

Can you create a separate ticket for that please?

**Comment by WinterSilence** [comment]
@orklah nope, in this case all fine - $baseUrl not defined in A and can't use `@inheritdoc` tag

**Comment by orklah** [comment]
I was talking about public properties being assigned from outside the class with incompatible type. I made a ticket here #6583

It will incidentally resolve your issue because Psalter will stop changing public properties, but your issue with inherited types will still be here on protected properties

**Comment by WinterSilence** [comment]
@orklah 1. Please fix example 2. I'm don't think what's "ignore public properties" is good behavior. I'm thing good behavior is skip properties with types in `@var`, they can be fix only optionally.

**Comment by orklah** [comment]
The issue I posted in #6583 can't be ignored. Psalter produce broken code.

About your suggestion, Psalter change properties with @var on purpose. It's to be able to add a signature type when possible.

The things to be done here are:
- make Psalm aware of inheritance on properties so that possible values in child are reflected in parent
- make Psalm inherit by default the type of the parent property (no need for @inheritDoc, Psalm works without it for methods)
- Probably have a special case for private properties as well

**Comment by WinterSilence** [comment]
@orklah 

> make Psalm aware of inheritance on properties so that possible values in child are reflected in parent

 ignore [LSP](https://en.wikipedia.org/wiki/Liskov_substitution_principle) is bad idea

> make Psalm inherit by default the type of the parent property (no need for @inheritdoc, Psalm works without it for methods)

again bad idea: property inherit description too + harder to update - `@inheritdoc` work as link: when you change type in parent class, than no need update child types.

I think need something like "risky rules" in [CS Fixer](https://github.com/FriendsOfPHP/PHP-CS-Fixer) or "interactive mode" to keep user's definitions.

**Comment by orklah** [comment]
It's not ignoring LSP. Actually property type are invariant. That means every type accepted by a child must be accepted in the parent and the other way around.

So if you put '...' in the child, the parent type should accept it.

**Comment by WinterSilence** [comment]
@orklah https://psalm.dev/r/4acd789e3e

**Comment by orklah** [comment]
Yeah, Psalm is emitting an issue because you don't have the same type in both, that's what I was saying.

**Comment by WinterSilence** [comment]
@orklah ok, but it's not same to "child are reflected in parent"

**Comment by orklah** [comment]
It totally is :) Iterable is wider than array so in this case, to fix the issue, you have to make the parent an iterable too. You can't have a child that contains type that are not accepted by the parent (and the other way around too)

**Comment by WinterSilence** [comment]
@orklah nope, `new B(new ArrayIterator([1, 2]))` work fine, but you can't `new A(new ArrayIterator([1, 2]))`

**Comment by orklah** [comment]
This would make no sense: https://psalm.dev/r/e7c3d3a7b3

What if I do `new B(new ArrayIterator([1, 2]))` and then I call `isArray()` on it? Psalm says it's always array because properties have to be invariant. Your example would not return true to isArray. You can't just make a class compatible with a wider type just by widening the property type

**Comment by WinterSilence** [comment]
@orklah 
> You can't just make a class compatible with a wider type just by widening the property type

I can - you don't understand LSP: you can extend type array -> iterable -> mixed, but you can't: mixed -> iterable -> array

**Comment by orklah** [comment]
Depends if you're talking about parameters, return types or properties.

Return types are covariant, you can always be more narrow than the parent
Parameter types are contravariant, you can always be wider than the parent
Property types are invariant, you can only have the same type as the parent

**Comment by WinterSilence** [comment]
@orklah now I'm understand you - `$items` don't have type in your example

**Comment by orklah** [comment]
Well it have types in phpdoc, but that's the same for Psalm.

But yeah, when they are in signature, PHP freaks out when they're not invariant:
https://3v4l.org/eTOo0

**Comment by WinterSilence** [comment]
@orklah 
>that's the same for Psalm

yes, but not correct in case as in my example when property type missed, but defined type of setter parameter
